File: frontend/src/test-utils/campaignScope.ts
export const CAMPAIGN_EXCLUSION_REASONS = [
  'parallel-scheduling',
  'parsl',
  'managed-remote',
  'distributed-engine',
] as const

export type CampaignExclusionReason = typeof CAMPAIGN_EXCLUSION_REASONS[number]
export const CAMPAIGN_ANNOTATION = 'campaign-excluded'
export const CAMPAIGN_META_KEY = 'campaignExcluded'

export interface CollectedCampaignCase {
  nodeid: string
  reason?: unknown
}

interface CampaignManifestSuite {
  runner: string
  root: string
  excluded: unknown[]
}

interface CampaignManifest {
  schema_version: number
  suites: Record<string, CampaignManifestSuite>
}

export function loadCampaignExclusions(
  document: unknown,
  suiteName: string,
  runner: string,
): Map<string, CampaignExclusionReason> {
  if (!document || typeof document !== 'object') throw new Error('campaign manifest must be an object')
  const manifest = document as Partial<CampaignManifest>
  if (manifest.schema_version !== 1) throw new Error('campaign manifest schema_version must be 1')
  const suite = manifest.suites?.[suiteName]
  if (!suite || suite.runner !== runner || suite.root !== 'frontend') {
    throw new Error(`${suiteName} suite must declare ${runner} with root 'frontend'`)
  }
  if (!Array.isArray(suite.excluded)) throw new Error(`${suiteName} excluded must be a list`)

  const exclusions = new Map<string, CampaignExclusionReason>()
  for (const [index, rawEntry] of suite.excluded.entries()) {
    if (!rawEntry || typeof rawEntry !== 'object') {
      throw new Error(`${suiteName} excluded entry ${index} must be an object`)
    }
    const entry = rawEntry as Record<string, unknown>
    if (Object.keys(entry).sort().join(',') !== 'nodeid,reason') {
      throw new Error(`${suiteName} excluded entry ${index} must contain only nodeid and reason`)
    }
    const { nodeid, reason } = entry
    if (typeof nodeid !== 'string' || !nodeid) {
      throw new Error(`${suiteName} excluded entry ${index} has an invalid nodeid`)
    }
    if (!CAMPAIGN_EXCLUSION_REASONS.includes(reason as CampaignExclusionReason)) {
      throw new Error(`${suiteName} excluded entry ${nodeid} has invalid reason ${String(reason)}`)
    }
    if (exclusions.has(nodeid)) {
      throw new Error(`campaign manifest has duplicate ${suiteName} nodeids: ${nodeid}`)
    }
    exclusions.set(nodeid, reason as CampaignExclusionReason)
  }
  return exclusions
}

export function auditCampaignCollection(
  suiteName: string,
  cases: CollectedCampaignCase[],
  manifest: Map<string, CampaignExclusionReason>,
): { selected: string[], excluded: string[] } {
  const errors: string[] = []
  const collected = new Set<string>()
  const marked = new Map<string, CampaignExclusionReason>()
  const duplicateNodeids = new Set<string>()
  for (const testCase of cases) {
    if (collected.has(testCase.nodeid)) duplicateNodeids.add(testCase.nodeid)
    collected.add(testCase.nodeid)
    if (testCase.reason === undefined) continue
    if (!CAMPAIGN_EXCLUSION_REASONS.includes(testCase.reason as CampaignExclusionReason)) {
      errors.push(`${testCase.nodeid}: invalid exclusion reason ${String(testCase.reason)}`)
      continue
    }
    const reason = testCase.reason as CampaignExclusionReason
    const previous = marked.get(testCase.nodeid)
    if (previous !== undefined && previous !== reason) {
      errors.push(`${testCase.nodeid}: conflicting collected exclusion reasons`)
      continue
    }
    marked.set(testCase.nodeid, reason)
  }
  if (duplicateNodeids.size) {
    errors.push(`duplicate collected nodeids: ${[...duplicateNodeids].sort().join(', ')}`)
  }
  const missing = [...manifest.keys()].filter(nodeid => !collected.has(nodeid)).sort()
  const unmanifested = [...marked.keys()].filter(nodeid => !manifest.has(nodeid)).sort()
  const unmarked = [...manifest.keys()].filter(nodeid => collected.has(nodeid) && !marked.has(nodeid)).sort()
  const drift = [...marked.entries()]
    .filter(([nodeid, reason]) => manifest.has(nodeid) && manifest.get(nodeid) !== reason)
    .map(([nodeid, reason]) => `${nodeid}: annotation=${reason}, manifest=${manifest.get(nodeid)}`)
    .sort()
  if (missing.length) errors.push(`manifest nodeids missing from collection: ${missing.join(', ')}`)
  if (unmanifested.length) errors.push(`marked nodeids missing from manifest: ${unmanifested.join(', ')}`)
  if (unmarked.length) errors.push(`manifest nodeids missing campaign annotations: ${unmarked.join(', ')}`)
  if (drift.length) errors.push(`campaign exclusion reason drift: ${drift.join(', ')}`)
  const selected = [...collected].filter(nodeid => !marked.has(nodeid)).sort()
  if (!selected.length) errors.push('campaign selection contains zero tests')
  if (errors.length) throw new Error(`invalid ${suiteName} local campaign scope:\n- ${errors.join('\n- ')}`)
  return { selected, excluded: [...marked.keys()].sort() }
}

File: backend/tests/campaign_scope.py
"""Strict manifest validation for the local-platform pytest campaign."""

from __future__ import annotations

from collections import Counter
import json
from pathlib import Path
from typing import Any

import pytest


ALLOWED_EXCLUSION_REASONS = frozenset(
    {"parallel-scheduling", "parsl", "managed-remote", "distributed-engine"}
)
MANIFEST_PATH = Path(__file__).parents[2] / "tests" / "campaign-local-scope.json"


def load_backend_exclusions(path: Path = MANIFEST_PATH) -> dict[str, str]:
    """Load the exact backend pytest exclusion map and reject malformed manifests."""

    try:
        document: Any = json.loads(path.read_text(encoding="utf-8"))
        if document.get("schema_version") != 1:
            raise ValueError("schema_version must be 1")
        suite = document["suites"]["backend"]
        if suite.get("runner") != "pytest" or suite.get("root") != "backend":
            raise ValueError("backend suite must declare pytest with root 'backend'")
        entries = suite["excluded"]
        if not isinstance(entries, list):
            raise ValueError("backend excluded must be a list")
    except (AttributeError, KeyError, OSError, TypeError, json.JSONDecodeError) as error:
        raise ValueError(f"cannot load campaign manifest {path}: {error}") from error

    nodeids: list[str] = []
    exclusions: dict[str, str] = {}
    for index, entry in enumerate(entries):
        if not isinstance(entry, dict) or set(entry) != {"nodeid", "reason"}:
            raise ValueError(
                f"backend excluded entry {index} must contain only nodeid and reason"
            )
        nodeid = entry["nodeid"]
        reason = entry["reason"]
        if not isinstance(nodeid, str) or not nodeid:
            raise ValueError(f"backend excluded entry {index} has an invalid nodeid")
        if reason not in ALLOWED_EXCLUSION_REASONS:
            raise ValueError(
                f"backend excluded entry {nodeid!r} has invalid reason {reason!r}"
            )
        nodeids.append(nodeid)
        exclusions[nodeid] = reason

    duplicates = sorted(nodeid for nodeid, count in Counter(nodeids).items() if count > 1)
    if duplicates:
        raise ValueError(f"campaign manifest has duplicate backend nodeids: {duplicates}")
    return exclusions


def audit_backend_collection(
    items: list[pytest.Item], manifest: dict[str, str]
) -> tuple[list[pytest.Item], list[pytest.Item]]:
    """Validate collected markers exactly, returning selected and excluded items."""

    errors: list[str] = []
    collected_ids = [item.nodeid for item in items]
    duplicate_collected = sorted(
        nodeid for nodeid, count in Counter(collected_ids).items() if count > 1
    )
    if duplicate_collected:
        errors.append(f"duplicate collected nodeids: {duplicate_collected}")

    marked: dict[str, str] = {}
    excluded: list[pytest.Item] = []
    for item in items:
        markers = list(item.iter_markers("campaign_excluded"))
        if not markers:
            continue
        if len(markers) != 1:
            errors.append(f"{item.nodeid}: expected one campaign_excluded marker")
            continue
        marker = markers[0]
        if marker.args or set(marker.kwargs) != {"reason"}:
            errors.append(
                f"{item.nodeid}: campaign_excluded requires only reason=<allowed reason>"
            )
            continue
        reason = marker.kwargs["reason"]
        if reason not in ALLOWED_EXCLUSION_REASONS:
            errors.append(f"{item.nodeid}: invalid exclusion reason {reason!r}")
            continue
        marked[item.nodeid] = reason
        excluded.append(item)

    collected = set(collected_ids)
    missing = sorted(set(manifest) - collected)
    unmanifested = sorted(set(marked) - set(manifest))
    unmarked = sorted(set(manifest) & collected - set(marked))
    drift = sorted(
        f"{nodeid}: marker={marked[nodeid]!r}, manifest={manifest[nodeid]!r}"
        for nodeid in set(marked) & set(manifest)
        if marked[nodeid] != manifest[nodeid]
    )
    if missing:
        errors.append(f"manifest nodeids missing from collection: {missing}")
    if unmanifested:
        errors.append(f"marked nodeids missing from manifest: {unmanifested}")
    if unmarked:
        errors.append(f"manifest nodeids missing campaign_excluded markers: {unmarked}")
    if drift:
        errors.append(f"campaign exclusion reason drift: {drift}")

    selected = [item for item in items if item.nodeid not in marked]
    if not selected:
        errors.append("campaign selection contains zero tests")
    if errors:
        raise ValueError("invalid backend local campaign scope:\n- " + "\n- ".join(errors))
    return selected, excluded
